The purpose of this r-markdown is to set up a program to run the r package "POWSC," created by Wu et al.


Install Software:
```{r results="hide"}
setRepositories(ind = 1:5)
install.packages(c("Biobase", "limma", "SingleCellExperiment", "MAST", "SummarizedExperiment"))

library(devtools)

install_github("haowulab/SC2P", build_vignettes=T)
install_github("suke18/POWSC", build_vignettes = T, dependencies = T)
```
**Note: SC2P is a package also create by the Wu Lab. It's purpose is to test DE for single cell RNA-seq data.


Estimation for a single cell type. In this case, we could upload a data set (here they use "es_mef_sce"). 

"sce" is defined by the data set. Specified to be looking at fibroblast cells. 

"est_Paras" is a list of data for of 44 cells (columns) by 11774 genes (rows). Each cell contains a value representing expression level. The list includes a matrix ($exprs), a vector of genes ($pi.g), a vector of cells ($p0), another vector for cells ($lambda), a vector for mu ($mu), a vector for standard deviation of each gene ($sd), and a vector sf for each cell ($sf).

"simData" is a simulated data set for 2 groups containing ix.DE1, ix.DE2, ix.DEs (which is a vector of ix.DE1 followed by ix.DE2), ix.DEGs (names to values of ix.DEs), pi.g1 and pig2 which seem to be normalized according to 3 RNA spike in's, mu1 and mu2, lfc, and pi.df. It comes from the function Simulate2SCE(). Simulate2SCE() simulates data for a 2 group comparrision. Multi group comparrision can be done using SimulateMultiSCEs.

"de" is list. It contains 3 tables for test type: cont, disc, hurdle; and 2 data frames containing 3 variables in each: geneIndex, pval, fdr.

"estPower" is a list that contains information on the power of the simulated experiemt. Here the author used Power_Disc and Power_Cont to list the power analyses. The difference between the two functions is the default delta value and strata.
```{r results="hide"}
library(POWSC)
data("es_mef_sce")
sce = es_mef_sce[, colData(es_mef_sce)$cellTypes == "fibro"]
est_Paras = Est2Phase(sce)

simData = Simulate2SCE(n=100, estParas1 = est_Paras, estParas2 = est_Paras)
de = runMAST(simData$sce)
estPower1 = Power_Disc(de, simData = simData)
estPower2 = Power_Cont(de, simData = simData)
```



Simulate data set and plot for two group comparison:

"runPOWSC" estimates power for a given data set. The input is an experimental size simulation, desired percent of differentially expressed genes, a template prarameter estiamted from a cell type, method of DE analysis (MAST or SC2P), a cell type, cell proportions, alpha, disc_delta which is the zero ratio changes cutoff, and cont_delta which is the log fold change cutoff.
```{r results="hide"}
sim_size = c(100, 400, 1000) # A numeric vector
pow_rslt = runPOWSC(sim_size = sim_size, est_Paras = est_Paras,per_DE=0.05, DE_Method = "MAST", Cell_Type = "PW") # Note, using our previous developed tool SC2P is faster.
#plot(pow_rslt, Form="I", Cell_Type = "Multi")
summary(pow_rslt,  Form="II", Cell_Type = "PW")
```

Upload GSE67835 for next simulation.
```{r results="hide"}

BiocManager::install("GEOquery")
library(GEOquery)
gse <- getGEO("GSE67835", GSEMatrix = TRUE)
show(gse)
```


Simulate data for multi-group comparison.
    (***Note: I substituted "sce" for "tmp_mat" on line 78. I also added a vector 1:44 for "sce$Patients", which was previously NULL)
    Plotting "pow_rslt" still a puzzle...for now
```{r results="hide"}
sim_size = 1000
cell_per = c(0.2, 0.3, 0.5)
data(gse$`GSE67835-GPL15520_series_matrix.txt.gz`)
col = colData(sce)
exprs = assays(sce)$counts
sce$Patients <- c(1:44)
(tb = table(colData(sce)$Patients, colData(sce)$cellTypes))
# use AB_S7 patient as example and take three cell types: astrocytes hybrid and neurons
estParas_set = NULL
celltypes = c("oligodendrocytes", "hybrid", "neurons")
for (cp in celltypes){
    print(cp)
    ix = intersect(grep(cp, col$cellTypes), grep("AB_S7", col$Patients))
    tmp_mat = exprs[, ix]
    tmp_paras = Est2Phase(sce)
    estParas_set[[cp]] = tmp_paras
}
pow_rslt = runPOWSC(sim_size = sim_size, est_Paras = estParas_set,per_DE=0.05, DE_Method = "MAST",multi_Prob = cell_per, Cell_Type = "Multi")
#plot(pow_rslt, Form="I", Cell_Type = "Multi")
summary(pow_rslt, Form="II", Cell_Type = "Multi")


```

